# Pandas - Cleaning Data

# Data Cleaning
Data cleaning means fixing bad data in your data set.

Bad data could be:

Empty cells
Data in wrong format
Wrong data
Duplicates
In this tutorial you will learn how to deal with all of them.

# Our Data Set
In the next chapters we will use this data set:

The data set contains some empty cells ("Date" in row 22, and "Calories" in row 18 and 28).

The data set contains wrong format ("Date" in row 26).

The data set contains wrong data ("Duration" in row 7).

The data set contains duplicates (row 11 and 12).

# Pandas - Cleaning Empty Cells


Empty Cells
Empty cells can potentially give you a wrong result when you analyze data.

Remove Rows
One way to deal with empty cells is to remove rows that contain empty cells.

This is usually OK, since data sets can be very big, and removing a few rows will not have a big impact on the result.

# Example
Return a new Data Frame with no empty cells:

In [1]:
import pandas as pd
df=pd.read_csv("example.csv")
df

Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14          14        60  '2020/12/14'    104       132     379.3
15          15        60  '2020/12/15'     98       123     275.0
16          16        60  '2020/12/16'     98       120     215.2
17          17        60  '2020/12/17'    100       120     300.0
18          18        45  '2020/12/18'     90       112       NaN
19          19        60  '2020/12/19'    103       123     323.0
20          20        45  '2020/12/20'     97       125     243.0
21          21        60  '2020/12/21'    108       131     364.2
22          22        45           NaN    100       119     282.0
23          23        60  '2020/12/23'    130       101     300.0
24          24        45  '2020/12/24'    105       132     246.0
25          25        60  '2020/12/25'    102       126     334.5
26          26        60    12/26/2020    100       120     250.0
27          27        60  '2020/12/27'     92       118     241.0
28          28        60  '2020/12/28'    103       132       NaN
29          29        60  '2020/12/29'    100       132     280.0
30          30        60  '2020/12/30'    102       129     380.3
31          31        60  '2020/12/31'     92       115     243.0

In [5]:
newdf=df.dropna()
print(newdf.to_string())

    Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14        

In [6]:
# Note: By default, the dropna() method returns a new DataFrame, and will not change the original.

In [7]:
# If you want to change the original DataFrame, use the inplace = True argument:

Example
Remove all rows with NULL values:

In [8]:
import pandas as pd

df = pd.read_csv('example.csv')

df.dropna(inplace = True)

print(df.to_string())

    Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14        

Note: Now, the dropna(inplace = True) will NOT return a new DataFrame, but it will remove all rows containg NULL values from the original DataFrame.

# Replace Empty Values
Another way of dealing with empty cells is to insert a new value instead.

This way you do not have to delete entire rows just because of some empty cells.

The fillna() method allows us to replace empty cells with a value:

Example
Replace NULL values with the number 130:

In [9]:
import pandas as pd

df = pd.read_csv('example.csv')

df.fillna(1300, inplace = True)

In [10]:
df

Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14          14        60  '2020/12/14'    104       132     379.3
15          15        60  '2020/12/15'     98       123     275.0
16          16        60  '2020/12/16'     98       120     215.2
17          17        60  '2020/12/17'    100       120     300.0
18          18        45  '2020/12/18'     90       112    1300.0
19          19        60  '2020/12/19'    103       123     323.0
20          20        45  '2020/12/20'     97       125     243.0
21          21        60  '2020/12/21'    108       131     364.2
22          22        45          1300    100       119     282.0
23          23        60  '2020/12/23'    130       101     300.0
24          24        45  '2020/12/24'    105       132     246.0
25          25        60  '2020/12/25'    102       126     334.5
26          26        60    12/26/2020    100       120     250.0
27          27        60  '2020/12/27'     92       118     241.0
28          28        60  '2020/12/28'    103       132    1300.0
29          29        60  '2020/12/29'    100       132     280.0
30          30        60  '2020/12/30'    102       129     380.3
31          31        60  '2020/12/31'     92       115     243.0

# Replace Using Mean, Median, or Mode
A common way to replace empty cells, is to calculate the mean, median or mode value of the column.

Pandas uses the mean() median() and mode() methods to calculate the respective values for a specified column:

Example
Calculate the MEAN, and replace any empty values with it:

In [12]:
# Mean = the average value (the sum of all values divided by number of values).
import pandas as pd

df = pd.read_csv('example.csv')

x = df["Calories"].median()

df["Calories"].fillna(x, inplace = True)

In [13]:
df

Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14          14        60  '2020/12/14'    104       132     379.3
15          15        60  '2020/12/15'     98       123     275.0
16          16        60  '2020/12/16'     98       120     215.2
17          17        60  '2020/12/17'    100       120     300.0
18          18        45  '2020/12/18'     90       112     291.2
19          19        60  '2020/12/19'    103       123     323.0
20          20        45  '2020/12/20'     97       125     243.0
21          21        60  '2020/12/21'    108       131     364.2
22          22        45           NaN    100       119     282.0
23          23        60  '2020/12/23'    130       101     300.0
24          24        45  '2020/12/24'    105       132     246.0
25          25        60  '2020/12/25'    102       126     334.5
26          26        60    12/26/2020    100       120     250.0
27          27        60  '2020/12/27'     92       118     241.0
28          28        60  '2020/12/28'    103       132     291.2
29          29        60  '2020/12/29'    100       132     280.0
30          30        60  '2020/12/30'    102       129     380.3
31          31        60  '2020/12/31'     92       115     243.0

In [14]:
# Median = the value in the middle, after you have sorted all values ascending.


# Example
Calculate the MODE, and replace any empty values with it:

In [15]:
import pandas as pd

df = pd.read_csv('example.csv')

x = df["Calories"].mode()[0]

df["Calories"].fillna(x, inplace = True)

In [16]:
df

Unnamed: 0  Duration          Date  Pulse  Maxpulse  Calories
0            0        60  '2020/12/01'    110       130     409.1
1            1        60  '2020/12/02'    117       145     479.0
2            2        60  '2020/12/03'    103       135     340.0
3            3        45  '2020/12/04'    109       175     282.4
4            4        45  '2020/12/05'    117       148     406.0
5            5        60  '2020/12/06'    102       127     300.0
6            6        60  '2020/12/07'    110       136     374.0
7            7       450  '2020/12/08'    104       134     253.3
8            8        30  '2020/12/09'    109       133     195.1
9            9        60  '2020/12/10'     98       124     269.0
10          10        60  '2020/12/11'    103       147     329.3
11          11        60  '2020/12/12'    100       120     250.7
12          12        60  '2020/12/12'    100       120     250.7
13          13        60  '2020/12/13'    106       128     345.3
14          14        60  '2020/12/14'    104       132     379.3
15          15        60  '2020/12/15'     98       123     275.0
16          16        60  '2020/12/16'     98       120     215.2
17          17        60  '2020/12/17'    100       120     300.0
18          18        45  '2020/12/18'     90       112     300.0
19          19        60  '2020/12/19'    103       123     323.0
20          20        45  '2020/12/20'     97       125     243.0
21          21        60  '2020/12/21'    108       131     364.2
22          22        45           NaN    100       119     282.0
23          23        60  '2020/12/23'    130       101     300.0
24          24        45  '2020/12/24'    105       132     246.0
25          25        60  '2020/12/25'    102       126     334.5
26          26        60    12/26/2020    100       120     250.0
27          27        60  '2020/12/27'     92       118     241.0
28          28        60  '2020/12/28'    103       132     300.0
29          29        60  '2020/12/29'    100       132     280.0
30          30        60  '2020/12/30'    102       129     380.3
31          31        60  '2020/12/31'     92       115     243.0